iroha_introspect/src/lib.rs

```diff
@@ -0,0 +1,358 @@
+//! Module for introspecting rust types in other languages for translation.
+
+use std::{
+    collections::{BTreeMap, BTreeSet},
+    time::Duration,
+};
+
+/// Derive introspect. It will make your structure introspectable
+pub use iroha_introspect_derive::Introspect;
+use serde::Serialize;
+
+/// Metadata map
+pub type MetaMap = BTreeMap<String, Metadata>;
+
+/// Introspect trait
+pub trait Introspect {
+    /// Returns unique type name.
+    /// WARN: `std::any::type_name` is compiler related, so is not unique.
+    /// I guess we should change it somehow later
+    fn type_name() -> String {
+        let mut name = module_path!().to_owned();
+        name.push_str("::");
+        name.push_str(std::any::type_name::<Self>());
+        name
+    }
+
+    /// Returns info about current type. Will return map from type names to its metadata
+    fn get_introspect() -> MetaMap {
+        let mut map = MetaMap::new();
+        Self::introspect(&mut map);
+        map
+    }
+
+    /// Introspect function. Give it empty map and it will return description of types
+    /// related to this type
+    fn introspect(metamap: &mut MetaMap);
+}
+
+/// Metadata
+#[derive(Debug, Eq, PartialEq, Hash, Ord, PartialOrd, Serialize, Clone)]
+pub enum Metadata {
+    /// Named structure
+    NamedStruct(NamedFieldsMeta),
+    /// Unnamed structure
+    UnnamedStruct(UnnamedFieldsMeta),
+    /// Enum
+    Enum(EnumMeta),
+    /// Integer
+    Int(IntMode),
+    /// String
+    String,
+    /// Bool
+    Bool,
+    /// Array
+    Array(ArrayMeta),
```

Comment:
Why do we need a separate Array type? I think it can be described as Vector

Reply:
It does make sence, to ensure contract (that it will always be of the same length)

Reply:
Ok, let it be